core/types/types.go

```diff
@@ -56,57 +50,35 @@ type BlockHeader struct {
 	StateRoot      common.Hash `json:"stateRoot"`
 	ExtrinsicsRoot common.Hash `json:"extrinsicsRoot"`
 	Digest         []byte      `json:"digest"` // any additional block info eg. logs, seal
+	hash           common.Hash
 }
 
-func (bh *BlockHeader) Hash() (common.Hash, error) {
-	enc, err := scale.Encode(bh)
-	if err != nil {
-		return [32]byte{}, err
-	}
+// SetHash sets a block header's hash. Note, you would probably want to use Hash() instead,
+// unless you know the hash has already been set
+func (bh *BlockHeader) SetHash(h common.Hash) {
+	bh.hash = h
+}
 
-	return common.Blake2bHash(enc)
+// GetHash retrieves the header's hash. Since it may be empty if not set, you probably want
+// to use Hash() instead.
+func (bh *BlockHeader) GetHash() common.Hash {
+	return bh.hash
 }
 
-func (bh *BlockHeader) WithHash() (*BlockHeaderWithHash, error) {
+// Hash hashes the block header, sets the header's internal hash field, and returns it
+func (bh *BlockHeader) Hash() (common.Hash, error) {
 	enc, err := scale.Encode(bh)
 	if err != nil {
-		return nil, err
+		return [32]byte{}, err
 	}
 
 	hash, err := common.Blake2bHash(enc)
 	if err != nil {
-		return nil, err
-	}
-
-	return &BlockHeaderWithHash{
-		ParentHash:     bh.ParentHash,
-		Number:         bh.Number,
-		StateRoot:      bh.StateRoot,
-		ExtrinsicsRoot: bh.ExtrinsicsRoot,
-		Digest:         bh.Digest,
-		Hash:           hash,
-	}, nil
-}
-
-func (bh *BlockHeaderWithHash) WithoutHash() *BlockHeader {
-	return &BlockHeader{
-		ParentHash:     bh.ParentHash,
-		Number:         bh.Number,
-		StateRoot:      bh.StateRoot,
-		ExtrinsicsRoot: bh.ExtrinsicsRoot,
-		Digest:         bh.Digest,
+		return [32]byte{}, err
 	}
-}
 
-// BlockHeaderWithHash is a state block header
-type BlockHeaderWithHash struct {
-	ParentHash     common.Hash `json:"parentHash"`
-	Number         *big.Int    `json:"number"`
-	StateRoot      common.Hash `json:"stateRoot"`
-	ExtrinsicsRoot common.Hash `json:"extrinsicsRoot"`
-	Digest         []byte      `json:"digest"` // any additional block info eg. logs, seal
-	// TODO: Not part of spec, can potentially remove
-	Hash common.Hash `json:"hash"`
+	bh.hash = hash
+	return hash, err
 }
```

Comment:
We don't need both of these. IMO the cleaner solution would be to always call `Hash` and we can implement caching later.

This might be a good time to think about how we are constructing blocks. The builder pattern seems to be rather common for this, using it might remove the need to track whether a hash is still valid as it would only be produced once the builder is done constructing the block. (maybe there's another way to enforce this?)

Reply:
yeah you're right, I have removed GetHash and added MustHash instead, also added NewBlockHeader which hashes the header for us. using NewBlockHeader only would be preferred then

Reply:
I don't think `MustHash` makes sense here. As we disccused yesterday it's maybe not the best choice to panic, also it doesn't solve the issue that you were trying to fix here (avoiding re-calculating the hash)

Reply:
Looking at the code I can see why it's so important to have a oneliner for getting the block hash. Maybe we can just call it `Hash` for now, and remove the panic later on. As I mentioned yesterday, I think a builder pattern would solve the issue of updating the hash correctly.

Reply:
okay, so should I leave the panic? alternatively we could just return an empty/zero hash if there's an error

Reply:
>  IMO we should just merge Hash and MustHash for now, and remove the use of panic later on when we can guarantee that the block is totally formed.

Reply:
okay, I have done that